### PR TITLE
[Bugfix] Scales between Levels - Reconstructed Initial Shape

### PR DESCRIPTION
**Pull request by nontas**
This PR introduces 3 major changes (*apologies for the long post :-)*):

#### 1) Bugfix
When fitting a multi-scale method, we would transform the fitting result of scale `i` into scale `i + 1` by doing:

```python
# Prepare this scale's final shape for the next scale
if i < self.n_scales - 1:
    # This should not be done for the last scale.
    shape = algorithm_result.final_shape   # final shape of scale i
    if self.scales[i] != self.scales[-1]:
        shape = Scale(self.scales[i + 1] / self.scales[i], 
                      n_dims=shape.n_dims).apply(shape)
```
This was wrong for the following reasons:
1. Everything was related to `self.scales[-1]`, thus in order for things to work the highest scale level should be equal to ``1.0``, i.e.  ``self.scales[-1] = 1.``.
2. Nothing worked in case the extracted `holistic_features()` introduced some kind of scaling transform (down-sampling). Features had to be *densely* sampled and return an image of the same shape as the original one.
3. Due to numerical issues, by applying `Scale(self.scales[i + 1] / self.scales[i])` was returning a slightly different shape (in terms of similarity transform; not deformations) than the original one. Thus the fitting chain was inaccurate since the initial shape of scale `i` was not *exactly* the same as the final shape of the previous scale `i - 1`.
4. The previous error means that the initial errors of our parametric methods were slightly wrong. The final errors were not, since in most cases (if not all) the maximum scale level was equal to ``1.0``.

All these issues are now fixed. When executing the `_prepare_image()` method during fitting, we now store the `affine_transforms` applied by rescaling to the reference shape and extracting features, as well as the `scale_transforms` applied by each level's scale factor. These transforms are then passed to the fitting result objects in order to accurately transform the `initial_shape`, `final_shape` and `shapes` into the original image's space. These transforms are also used in order to correctly transform the shape of scale `i` to scale `i + 1` during fitting, using the following code (it replaces the previews code logic):

```python
# Prepare this scale's final shape for the next scale
if i < self.n_scales - 1:
    # This should not be done for the last scale.
    shape = algorithm_result.final_shape   # final shape of scale i
    if self.holistic_features[i + 1] != self.holistic_features[i]:
        # If the features function of the current scale is different
        # than the one of the next scale, this means that the affine
        # transform is different as well. Thus we need to do the
        # following composition:
         #
         #    S_{i+1} \circ A_{i+1} \circ inv(A_i) \circ inv(S_i)
         #
        # where:
        #    S_i : scaling transform of current scale
        #    S_{i+1} : scaling transform of next scale
        #    A_i : affine transform of current scale
        #    A_{i+1} : affine transform of next scale
        t1 = scale_transforms[i].compose_after(affine_transforms[i])
        t2 = affine_transforms[i + 1].pseudoinverse().compose_after(t1)
        transform = scale_transforms[i + 1].pseudoinverse().compose_after(t2)
        shape = transform.apply(shape)
    elif (self.holistic_features[i + 1] == self.holistic_features[i]
              and self.scales[i] != self.scales[i + 1]):
        # If the features function of the current scale is the same
        # as the one of the next scale, this means that the affine
        # transform is the same as well, and thus can be omitted.
        # Given that the scale factors are different, we need to do
        # the # following composition:
        #
        #    S_{i+1} \circ inv(S_i)
        #
        # where:
        #    S_i : scaling transform of current scale
        #    S_{i+1} : scaling transform of next scale
        transform = scale_transforms[i + 1].pseudoinverse().compose_after(
                scale_transforms[i])
        shape = transform.apply(shape)
```

#### 2) Reconstructed Initial Shapes
When fitting a parametric method, i.e. perform some kind of optimisation over the parameters of a statistical shape model, the provided `initial_shape` first gets reconstructed using the shape model. This involves projected the shape on the basis in order to get the initial value of the shape parameters and then reconstructing it in order to create the initial shape of the optimisation. Until now, the fitting result object of such a fitting procedure would return the reconstructed shape as `initial_shape`, which makes sense since it is the initial point of the performed optimisation. Due to relevant requests, this PR makes the above procedure more clear. A `MultiScaleParametricIterativeResult` now stores the shape *provided by the user* as `initial_shape` and its reconstruction as `reconstructed_initial_shape`. It also clearly documents the above behaviour where applicable. A scenario where this change is useful is the following:

> If I attempt to perform a fitting with the initial shape being the ground truth shape, then, until now, the initial error would not be ``0.`` as expected because it was based on the initial shape's reconstruction. The changes of this PR would return `result.initial_error() == 0.` and `result.reconstructed_initial_error() != 0.`.

Note that this is done for the initial shapes of all scales of a fitting. Thus, this PR correctly sets the corresponding legend entries of `result.view_iterations()` to ``Reconstruction``, since those reconstruction steps are not considered as separate iterations and do not count on `n_iters`.

#### 3) *Fitter* Objects
This PR also slightly changes the abstract structure of `Fitter` objects in order to be more clear and in correspondence to our `Result` objects. The old `MultiFitter` and `ModelFitter` are now replaced by `MultiScaleNonParametricFitter` and `MultiScaleParametricFitter`, respectively. Also, they are no longer abstract classes, but both have a constructor with the following signature:

```python
class MultiScaleNonParametricFitter(object):
    def __init__(self, scales, reference_shape, holistic_features, algorithms):
        self._scales = scales
        self._reference_shape = reference_shape
        self._holistic_features = holistic_features
        self.algorithms = algorithms
```
This is because all our methods require these properties.

It would be very useful if someone went through this PR to make sure everything is fine, because they are quite sensitive changes and we don't want anything to be broken towards the new release. We could also run some AAM and SDM experiments to see any performance difference.

### CODE REVIEW

**Comment by nontas** [comment]
I also forgot to mention above that I removed the `perturb_from_gt_bb()` and `perturb_from_bb()` methods for (the old) `ModelFitter` class. They were static methods that do not really make much sense, since they were directly calling `menpofit.fitter.noisy_shape_from_shape()`.

**Comment by nontas** [comment]
`menpowidgets` are updated in https://github.com/menpo/menpowidgets/pull/10. The changes refer to the `visualize_fitting_result` widget mainly.

I'm not sure why AppVeyor fails.

**Comment by nontas** [comment]
I ran some experiments that compare the behaviour of `menpofit` techniques with the respect to these changes.

## SDM + AFW
sdm_afw
<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th></th>      <th>mean</th>      <th>std</th>      <th>median</th>      <th>mad</th>      <th>max</th>      <th>auc</th>      <th>fr</th>    </tr>  </thead>  <tbody>    <tr>      <th>initial_old</th>      <td>0.1215</td>      <td>0.0704</td>      <td>0.1119</td>      <td>0.0375</td>      <td>0.8027</td>      <td>0.0300</td>      <td>0.8635</td>    </tr>    <tr>      <th>initial_new</th>      <td>0.1209</td>      <td>0.0704</td>      <td>0.1108</td>      <td>0.0374</td>      <td>0.8007</td>      <td>0.0310</td>      <td>0.8576</td>    </tr>    <tr>      <th>final_old</th>      <td>0.0335</td>      <td>0.0489</td>      <td>0.0236</td>      <td>0.0053</td>      <td>0.7169</td>      <td>0.5389</td>      <td>0.0564</td>    </tr>    <tr>      <th>final_new</th>      <td>0.0334</td>      <td>0.0479</td>      <td>0.0232</td>      <td>0.0055</td>      <td>0.6991</td>      <td>0.5400</td>      <td>0.0593</td>    </tr>  </tbody></table>

## SDM + LFPW testset
sdm_lfpw
<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th></th>      <th>mean</th>      <th>std</th>      <th>median</th>      <th>mad</th>      <th>max</th>      <th>auc</th>      <th>fr</th>    </tr>  </thead>  <tbody>    <tr>      <th>initial_old</th>      <td>0.0986</td>      <td>0.0489</td>      <td>0.0855</td>      <td>0.0293</td>      <td>0.3160</td>      <td>0.0448</td>      <td>0.7857</td>    </tr>    <tr>      <th>initial_new</th>      <td>0.0980</td>      <td>0.0489</td>      <td>0.0849</td>      <td>0.0291</td>      <td>0.3132</td>      <td>0.0469</td>      <td>0.7768</td>    </tr>    <tr>      <th>final_old</th>      <td>0.0229</td>      <td>0.0120</td>      <td>0.0202</td>      <td>0.0044</td>      <td>0.1329</td>      <td>0.6276</td>      <td>0.0179</td>    </tr>    <tr>      <th>final_new</th>      <td>0.0226</td>      <td>0.0121</td>      <td>0.0195</td>      <td>0.0039</td>      <td>0.1335</td>      <td>0.6313</td>      <td>0.0179</td>    </tr>  </tbody></table>

So it seems that we get a (even a marginal) improvement. I'll post AAM results as well as soon as they finish.

**Comment by nontas** [comment]
## PatchAAM + AFW
aam_afw
<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th></th>      <th>mean</th>      <th>std</th>      <th>median</th>      <th>mad</th>      <th>max</th>      <th>AUC</th>      <th>Failure Rate</th>    </tr>  </thead>  <tbody>    <tr>      <th>initial_old</th>      <td>0.1209</td>      <td>0.0700</td>      <td>0.1112</td>      <td>0.0365</td>      <td>0.7951</td>      <td>0.0304</td>      <td>0.8576</td>    </tr>    <tr>      <th>initial_new</th>      <td>0.1203</td>      <td>0.0701</td>      <td>0.1106</td>      <td>0.0369</td>      <td>0.7972</td>      <td>0.0309</td>      <td>0.8516</td>    </tr>    <tr>      <th>final_old</th>      <td>0.0545</td>      <td>0.0753</td>      <td>0.0242</td>      <td>0.0064</td>      <td>0.5887</td>      <td>0.4858</td>      <td>0.1780</td>    </tr>    <tr>      <th>final_new</th>      <td>0.0547</td>      <td>0.0755</td>      <td>0.0241</td>      <td>0.0064</td>      <td>0.5956</td>      <td>0.4855</td>      <td>0.1780</td>    </tr>  </tbody></table>

## PatchAAM + LFPW testset
aam_lfpw
<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th></th>      <th>mean</th>      <th>std</th>      <th>median</th>      <th>mad</th>      <th>max</th>      <th>auc</th>      <th>fr</th>    </tr>  </thead>  <tbody>    <tr>      <th>initial_old</th>      <td>0.0976</td>      <td>0.0488</td>      <td>0.0847</td>      <td>0.0294</td>      <td>0.3134</td>      <td>0.0471</td>      <td>0.7679</td>    </tr>    <tr>      <th>initial_new</th>      <td>0.0970</td>      <td>0.0487</td>      <td>0.0859</td>      <td>0.0289</td>      <td>0.3096</td>      <td>0.0487</td>      <td>0.7589</td>    </tr>    <tr>      <th>final_old</th>      <td>0.0291</td>      <td>0.0347</td>      <td>0.0212</td>      <td>0.0044</td>      <td>0.2802</td>      <td>0.5979</td>      <td>0.0446</td>    </tr>    <tr>      <th>final_new</th>      <td>0.0292</td>      <td>0.0344</td>      <td>0.0211</td>      <td>0.0044</td>      <td>0.2778</td>      <td>0.5956</td>      <td>0.0491</td>    </tr>  </tbody></table>

**Comment by patricksnape** [comment]
Right so it's basically identical - but now correct. Happy with that - are you still seeing the SDM problem with the keys?

**Comment by nontas** [comment]
Yes I still see the error. The following code:

```python
import menpo.io as mio
from menpofit.sdm import SupervisedDescentFitter, NonParametricNewton

path_to_images = '/home/nontas/Dropbox/lfpw/trainset/'

images = []
for im in mio.import_images(path_to_images, max_images=30, verbose=True):
    im.landmarks['bbox'] = im.landmarks['PTS'].lms.bounding_box()
    images.append(im)
    
sdm = SupervisedDescentFitter(images, group='PTS', bounding_box_group_glob='bbox', 
                              sd_algorithm_cls=NonParametricNewton, verbose=True)
```
gives this output:
```
Found 30 assets, index the returned LazyList to import.
- Computing reference shape                                                     Computing batch 0
- Normalizing images size: [==========] 100% (30/30) - done.                    
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-1-40b68b8b0913> in <module>()
     10 
     11 sdm = SupervisedDescentFitter(images, group='PTS', bounding_box_group_glob='bbox', 
---> 12                               sd_algorithm_cls=NonParametricNewton, verbose=True)

/home/nontas/Documents/Research/menpofit/menpofit/sdm/fitter.py in __init__(self, images, group, bounding_box_group_glob, sd_algorithm_cls, reference_shape, diagonal, holistic_features, patch_features, patch_shape, scales, n_iterations, n_perturbations, perturb_from_gt_bounding_box, batch_size, verbose)
    186         self._train(images, increment=False,  group=group,
    187                     bounding_box_group_glob=bounding_box_group_glob,
--> 188                     verbose=verbose, batch_size=batch_size)
    189 
    190     def _setup_algorithms(self):

/home/nontas/Documents/Research/menpofit/menpofit/sdm/fitter.py in _train(self, images, increment, group, bounding_box_group_glob, verbose, batch_size)
    237                 image_batch, increment=increment, group=group,
    238                 bounding_box_group_glob=bounding_box_group_glob,
--> 239                 verbose=verbose)
    240 
    241     def _train_batch(self, image_batch, increment=False, group=None,

/home/nontas/Documents/Research/menpofit/menpofit/sdm/fitter.py in _train_batch(self, image_batch, increment, group, bounding_box_group_glob, verbose)
    252             image_batch, self.n_perturbations,
    253             self._perturb_from_gt_bounding_box, gt_group=group,
--> 254             bb_group_glob=bounding_box_group_glob, verbose=verbose)
    255 
    256         # For each scale (low --> high)

/home/nontas/Documents/Research/menpofit/menpofit/fitter.py in generate_perturbations_from_gt(images, n_perturbations, perturb_func, gt_group, bb_group_glob, verbose)
    767 
    768         k = 0
--> 769         for bb in bb_generator(im):
    770             for _ in range(n_perturbations):
    771                 p_s = perturb_func(gt_s, bb).bounding_box()

/home/nontas/Documents/Research/menpofit/menpofit/fitter.py in bb_glob(im)
    742     else:
    743         def bb_glob(im):
--> 744             for k, v in im.landmarks.items_matching(bb_group_glob):
    745                 yield v.lms.bounding_box()
    746         bb_generator = bb_glob

/home/nontas/Documents/Research/menpo/menpo/landmark/base.py in items_matching(self, glob_pattern)
    280             Tuple of group, LandmarkGroup where the group matches the glob
    281         """
--> 282         for k, v in self.items():
    283             if fnmatch.fnmatch(k, glob_pattern):
    284                 yield k, v

/home/nontas/miniconda2/envs/menpo3/lib/python3.4/_collections_abc.py in __iter__(self)
    495 
    496     def __iter__(self):
--> 497         for key in self._mapping:
    498             yield (key, self._mapping[key])
    499 

RuntimeError: dictionary changed size during iteration
```

This is only happening for `Python 3.4`. It works fine on `Python 2.7`.

**Comment by patricksnape** [comment]
```
    if bb_group_glob is None:
        bb_generator = lambda im: [im.landmarks[gt_group].lms.bounding_box()]
        n_bbs = 1
    else:
        def bb_glob(im):
            return [v.lms.bounding_box()
                    for _, v in im.landmarks.items_matching(bb_group_glob)]
        bb_generator = bb_glob
        n_bbs = len(bb_glob(images[0]))
```

Can't push to your menpofit! Change `menpofit/fitter.py L739` to the above.

**Comment by nontas** [comment]
Thanks @patricksnape! That fixes the issue! Btw you are in the list of collaborators for my repo, so you should be able to push directly.

**Comment by patricksnape** [comment]
I can confirm that the SDM seems to be working for me - using my old scripts. Can't remember what the error is supposed to be but that is 10% than @nontas which seems pretty good.
selection_028

**Comment by jalabort** [comment]
Why is there such a big difference in terms of sdm performance between your results @patricksnape and @nontas? Are them on the same db?

**Comment by patricksnape** [comment]
We used different error metrics :laughing: But I think those results were consistent with what I was seeing before. They are trained on LFPW+HELEN rotated 4 times, same parameters as @nontas for features and ridge regression of 1000. I tried no ridge, was similar as above but with `60% @ 0.02`. This should be directly comparable to Yorgos graphs from the PO-CR paper.

**Comment by jalabort** [comment]
Right, makes sense now.

File: menpofit/dlib/fitter.py [hunk 1]
```diff
@@ -135,32 +137,39 @@ def __init__(self, images, group=None, bounding_box_group_glob=None,
...
+        # Call superclass
+        super(DlibERT, self).__init__(
+            scales=scales, reference_shape=reference_shape,
+            holistic_features=holistic_features, algorithms=[])
+
+        # Set parameters
+        self.diagonal = diagonal
         self.n_perturbations = n_perturbations
-        self.n_iterations = checks.check_max_iters(n_iterations, self.n_scales)
+        self.n_iterations = checks.check_max_iters(n_iterations, len(scales))
```
Comment (patricksnape) [inline menpofit/dlib/fitter.py]: Use `n_scales` from above?

File: menpofit/fitter.py [hunk 2]
```diff
@@ -399,89 +463,250 @@ def _fit(self, images, initial_shape, gt_shapes=None, max_iters=20,
...
+                    # following composition:
+                    #
+                    #    S_{i+1} \circ A_{i+1} \circ inv(A_i) \circ inv(S_i)
+                    #
+                    # where:
+                    #    S_i : scaling transform of current scale
+                    #    S_{i+1} : scaling transform of next scale
+                    #    A_i : affine transform of current scale
+                    #    A_{i+1} : affine transform of next scale
+                    t1 = \
```
Comment (patricksnape) [inline menpofit/fitter.py]: Let these go over 80 chars - this is confusing looking.

File: menpofit/sdm/fitter.py [hunk 3]
```diff
@@ -308,14 +336,36 @@ def _train_batch(self, image_batch, increment=False, group=None,
...
+                    # them.
+                    for jj, image_shapes in enumerate(current_shapes):
+                        for k, shape in enumerate(image_shapes):
+                            c_key = '__sdm_current_shape_{}'.format(k)
+                            feature_images[jj].landmarks[c_key] = \
+                                scale_transforms[jj].apply(shape)
+            else:
+                # Check if original training image (image_batch) got some current
+                # shape estimations attached. If yes, delete them.
+                if '__sdm_current_shape_0' in image_batch[0].landmarks.keys():
```
Comment (patricksnape) [inline menpofit/sdm/fitter.py]: Here you should be able to just do `if '__sdm_current_shape_0' in image_batch[0].landmarks`

**Comment by patricksnape** [comment]
Ok I just had a couple of comments about style. Other than that, the SDM appears to work for me and the AAM clearly works given your tests so +1.

Though I would wait a while before we release so we can test everything more thoroughly.

**Comment by nontas** [comment]
Thanks @patricksnape ! This is now ready. I agree we should test it more. It can be done while updating `menpofit-notebooks`.

**Pull request merged**